### PR TITLE
dataflow: even more refactors on reduce rendering

### PR DESCRIPTION
**Pull request by ruchirK**
Touches #4991

This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5002)

### CODE REVIEW

File: src/dataflow/src/render/reduce.rs [hunk 1]
```diff
@@ -187,163 +194,113 @@ where
...
-                                // should all be sorted before any unfused aggregates, so that we can grab them efficiently and at the
-                                // extract the subsets of inputs that corresponds to only unfused aggregates.
-                                for i in 0..input.len() {
-                                    match (input[i].0).0 {
-                                        ReductionType::FusedAccumulable => {
-                                            accumulable = (input[i].0).2.iter();
-                                            input_to_skip = Some(i);
+                                let mut basic = DatumList::empty().iter();
+
+                                // TODO assert that there are <= 3 items?
```
Comment (ruchirK) [inline src/dataflow/src/render/reduce.rs:273]: would appreciate second opinions on this

**Comment by ruchirK** [comment]
I think this is ready for a look.

I think its an open question of how much of this decision making we want do upstream but regardless I think this structure simplifies the logic in this file a fair amount. I hope others think so as well.

**Comment by ruchirK** [comment]
Strange. It looks like I'm getting a failure in kinesis.td before we even touch Materialize?

File: src/dataflow/src/render/reduce.rs [hunk 2]
```diff
@@ -361,17 +318,58 @@ where
     }
 }
 
+fn build_basic<G>(
```
Comment (ruchirK) [inline src/dataflow/src/render/reduce.rs:321]: lets add a test with a query + 2 or more jsonb_aggs

File: src/dataflow/src/render/reduce.rs [hunk 3]
```diff
@@ -28,20 +31,26 @@ use repr::{Datum, DatumList, Row, RowArena, RowPacker};
...
 enum ReductionType {
-    FusedAccumulable = 1,
-    FusedHierarchical = 2,
-    Unfused = 3,
+    // Accumulable functions can be subtracted from (are invertible), and associative.
+    // We can compute these results by moving some data to the diff field under arbitrary
+    // changes to inputs. Examples include sum or count.
+    Accumulable = 1,
+    // Hierarchical functions are associative, which means we can split up the work of
+    // computing them across subsets. Examples include min or max.
```
Comment (frankmcsherry) [inline src/dataflow/src/render/reduce.rs:42]: Minor, but I think an important-ish characteristic is that in addition to being associative, we expect data reduction when the function is called. For example, a hypothetical `sorted_agg` function could be made associative, but there wouldn't be a reason to make it hierarchical. Possibly out of scope for this moment in the code (though perhaps these tags really mean to signify which implementation we'll use).

File: src/dataflow/src/render/reduce.rs [hunk 4]
```diff
@@ -187,163 +194,113 @@ where
...
-
-                    if is_accumulable {
-                        accumulable.push((index, aggregates[index].clone()));
-                    } else if is_hierarchical {
-                        hierarchical.push((index, aggregates[index].clone()));
-                    } else {
-                        remaining.push((index, aggregates[index].clone()));
-                    }
+                    let typ = reduction_type(&aggregates[index].func);
+                    let aggregates_list = reduction_types.entry(typ).or_insert(Vec::new());
```
[the remark this answers is not in the document]
Comment (ruchirK) [reply src/dataflow/src/render/reduce.rs:202]: it didn't? I've written code like this before and clippy seems fine with it

File: src/dataflow/src/render/reduce.rs (hunk 3 above)
Comment (ruchirK) [reply src/dataflow/src/render/reduce.rs:42]: I think I agree with that. I can edit the comment later

File: src/dataflow/src/render/reduce.rs (hunk 1 above)
Comment (ruchirK) [reply src/dataflow/src/render/reduce.rs]: Ok so I decided to add a log line when we get negative multiplicities in the input vector -- I'm not 100% sure if thats right @frankmcsherry wdyt?

**Pull request merged**